webapi/src/main/scala/org/knora/webapi/messages/util/search/gravsearch/types/InferringGravsearchTypeInspector.scala

```diff
@@ -241,70 +241,110 @@ class InferringGravsearchTypeInspector(nextInspector: Option[GravsearchTypeInspe
-                                    Set.empty[GravsearchEntityTypeInfo]
-                            }
+                    val typesInferredFromPropertyIris: Set[GravsearchEntityTypeInfo] = inferFromStatements(statementsWithPropertyIris)
+                    val typesInferredFromVariablesAsPredicates: Set[GravsearchEntityTypeInfo] = inferFromStatements(statementsWithVariablesAsPredicates)
+
+                    // If any types were inferred from statements whose predicates are IRIs, update the
+                    // intermediate type inspection result with that information.
+
+                    val intermediateResultWithTypesInferredFromPropertyIris = if (typesInferredFromPropertyIris.nonEmpty) {
+                        updatedIntermediateResult.addTypes(entityToType, typesInferredFromPropertyIris, inferredFromPropertyIri = true)
+                    } else {
+                        updatedIntermediateResult
                     }
 
+                    val intermediateResultWithTypesInferredFromVariablesAsPredicates = intermediateResultWithTypesInferredFromPropertyIris.addTypes(
+                        entityToType,
+                        typesInferredFromVariablesAsPredicates
+                    )
+
+                    runNextRule(
+                        entityToType = entityToType,
+                        intermediateResult = intermediateResultWithTypesInferredFromVariablesAsPredicates,
+                        entityInfo = entityInfo,
+                        usageIndex = usageIndex
+                    )
+
                 case None =>
                     // This entity hasn't been used as a statement object, so this rule isn't relevant.
-                    Set.empty[GravsearchEntityTypeInfo]
+                    runNextRule(
+                        entityToType = entityToType,
+                        intermediateResult = updatedIntermediateResult,
+                        entityInfo = entityInfo,
+                        usageIndex = usageIndex
+                    )
             }
-
-            runNextRule(
-                entityToType = entityToType,
-                intermediateResult = updatedIntermediateResult.addTypes(entityToType, inferredTypes, inferredFromProperty = true),
-                entityInfo = entityInfo,
-                usageIndex = usageIndex
-            )
         }
     }
 
-
     /**
-     * Infers an entity's type if the entity is used as the subject of a statement and
-     * the predicate's `knora-api:subjectType` is known.
+     * Infers an entity's type if the entity is used as the subject of a statement in which the predicate is a
+     * property IRI whose `knora-api:subjectType` is known.
      */
-    private class TypeOfSubjectFromPropertyRule(nextRule: Option[InferenceRule]) extends InferenceRule(nextRule = nextRule) {
+    private class InferTypeOfSubjectFromPredicateIri(nextRule: Option[InferenceRule]) extends InferenceRule(nextRule = nextRule) {
```

Comment:
I think `SubjectType` is ambiguous: it could mean "infer the expected subject type of the property" or "infer the type of an entity that is used as the subject of a property". But `TypeOfSubject` makes it clear that we're inferring information about a subject, not about a property.